v2/cmd/asoctl/internal/importing/find_resources_by_scope.go

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package importing
+
+import (
+	"sync"
+
+	"github.com/Azure/azure-service-operator/v2/api"
+	"github.com/Azure/azure-service-operator/v2/internal/set"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+var (
+	resourceTypesByScope         map[genruntime.ResourceScope]set.Set[string]
+	populateResourceTypesByScope sync.Once
+)
+
+// FindResourceTypesByScope returns the resource types that may be parented in the given scope
+func FindResourceTypesByScope(scope genruntime.ResourceScope) []string {
+	ensureResourceTypesByScope()
+	s, ok := resourceTypesByScope[scope]
+	if !ok {
+		// No resource types for that scope
+		return nil
+	}
+
+	return s.Values()
+}
+
+// IsExtensionType returns true if the given type name is an extension type
+func IsExtensionType(typeName string) bool {
+	ensureResourceTypesByScope()
+	s, ok := resourceTypesByScope[genruntime.ResourceScopeExtension]
+	if !ok {
+		// No known extension resource types
+		return false
+
+	}
+
+	return s.Contains(typeName)
+}
+
+// ensureResourceTypesByScope ensures that the resourceTypesByScope map is populated
+func ensureResourceTypesByScope() {
+	populateResourceTypesByScope.Do(func() {
+		resourceTypesByScope = createResourceTypesByScope()
+	})
+}
+
+func createResourceTypesByScope() map[genruntime.ResourceScope]set.Set[string] {
+	result := make(map[genruntime.ResourceScope]set.Set[string], 4)
```

Comment:
minor: You don't need to specify 4 here, the linter is fine without it. I know 4 is technically correct (as there are 4 possible scopes), but in the absence of a way to actually pin the initial size of this collection to the number of supported scopes (there's no `len(enum)` unfortunately) it feels like it would be better to just not specify an initial size at all?

Reply:
I added this line _because the linter complained_.
It feels very nondeterministic. 
I'll remove the size and wait for it to complain.

Reply:
I believe the rule is that if the loop has no returns or breaks, it wants you to prealloc it. If there _are_ returns or breaks such that it's not obvious what the actual resultant count will be, you don't need to prealloc.

See: https://github.com/alexkohler/prealloc#flags